src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_IndexerAccess.cs

```diff
@@ -265,46 +293,130 @@ private BoundExpression MakePatternIndexOffsetExpression(BoundExpression unlower
             MethodSymbol sliceMethod,
             BoundExpression rangeArg)
         {
-            // Lowered code:
+            Debug.Assert(TypeSymbol.Equals(
+                rangeArg.Type,
+                _compilation.GetWellKnownType(WellKnownType.System_Range),
+                TypeCompareKind.ConsiderEverything));
+
+            // Lowered code without optimizations:
             // var receiver = receiverExpr;
             // int length = receiver.length;
             // Range range = argumentExpr;
             // int start = range.Start.GetOffset(length)
-            // int end = range.End.GetOffset(length)
-            // receiver.Slice(start, end - start)
+            // int rangeSize = range.End.GetOffset(length) - start
+            // receiver.Slice(start, rangeSize)
 
             var F = _factory;
 
+            var localsBuilder = ArrayBuilder<LocalSymbol>.GetInstance();
```

Comment:
>.GetInstance(); [](start = 57, length = 15)

On line 221, you use an optimistic size (3), but here and on line below you don't pass any size. Is that worth aligning?

Reply:
I'm making a very subtle guess here. For Index, there's a pattern optimization where we don't need the Length if the Index is from the start, being created inline via a conversion. So the code would have to look like:

```C#
int offset = ...;
int item = collection[(Index)offset];
```

Note that this won't work if the offset is either from the end, or if there's no Index conversion (since the normal behavior is to call the indexer that takes an `int` directly).

So my prediction is that almost every pattern Index indexer optimization will be from the end. Thus, I hint that explicitly.

For Range, it's completely different. Code like `0..5` seems like it will very common, and because Range only takes Indexes, the construction requires the use of the Index -> int explicit conversion, that I try to remove optimistically. So in this case, if the user is making a range only from start, or only from end, they won't or will require the length, respectively.

Here, I have no idea which will be more common, so I don't hint anything.

Do you think this is worth a comment?

Reply:
Fine to leave as-is. Thanks

---
In reply to: [291857095](https://github.com/dotnet/roslyn/pull/36147#discussion_r291857095) [](ancestors = 291857095)